spmd/tensor/ops/tp_sharding_ops.py

```diff
@@ -4,40 +4,36 @@
 import torch.utils._pytree as pytree
 from typing import List
 from spmd.tensor.api import DTensor
+from spmd.tensor.dispatch import OpSchema, OutputSharding
 from spmd.tensor.utils import unwrap_local_tensor
-from spmd.tensor.ops.utils import unwrap_single_placement, register_impl
+from spmd.tensor.ops.utils import unwrap_single_placement, register_impl, register_prop_rule
+from spmd.tensor.placement_types import DTensorSpec
 
 """
 The ops below were quickly hacked and needed to be polished down the road.
 Although they come with unit tests already, the logic is directly borrowed
-from ShardedTensor. We need to also make it work for all placement types 
+from ShardedTensor. We need to also make it work for all placement types
 of DTensor and all corner cases for sharded distributed tensor.
 """
 
 
-@register_impl("aten.baddbmm.default")
-def dist_baddbmm(
-    self: DTensor,
-    batch1: DTensor,
-    batch2: DTensor,
-    beta: float = 1.0,
-    alpha: float = 1.0,
-) -> DTensor:
-    local_input, local_batch1, local_batch2 = pytree.tree_map(
-        unwrap_local_tensor, (self, batch1, batch2)
+@register_prop_rule("aten.bmm.default")
+def bmm_rules(op_schema: OpSchema) -> OutputSharding:
+    mat1_spec, mat2_spec = op_schema.args_spec
+    mat1_shape = list(mat1_spec.shape) # b * n * m
+    mat2_shape = list(mat2_spec.shape) # b * m * p
+    return OutputSharding(
+        DTensorSpec(
+            mat1_spec.mesh,
+            mat1_spec.placements,
+            shape=list([mat1_shape[0], mat1_shape[1], mat2_shape[-1]]) # b * n * p
```

Comment:
I actually don't think we need to specify the output shapes (the shape field is purely there used by inputs and sharding propagation rule, that's why it's optional), the output shapes should be auto calculated when generating the output DTensor, so the shape field can leave as empty.

Reply:
I see. So the original assignment `shape=mat1_spec.shape` is only to initialize `self.ndim` in `__post_init__` function? 

"the output shapes should be auto calculated when generating the output DTensor, so the shape field can leave as empty."
> That's sweet. Then assigning an `ndim` parameter seems more straightforward IMO because it implies that the output shape is auto calculated.

Reply:
Oh hmmm yeah it did get assigned in some ops. I guess we can also assign it in this case then. My point is that this shouldn't be the burden of op implementation (i.e. we shouldn't enforce it), but it's ok to add if it's easy to tell the output shape

Reply:
I agree with you. By the way, could you share with a code pointer to where the shape information gets auto calculated? I did some search but did not find the place. Just to make sure that it happens somewhere.

Reply:
It's getting recovered during DTensor construction https://github.com/pytorch/PiPPy/blob/main/spmd/tensor/api.py#L147

Reply:
Just to note that for tracing mode, we won't have that -- but we should be able to use other types of shape propagation. @wanchaol  do we have a plan for that? Would meta tensors work here?

Reply:
@aazzolini I think for tracing mode we should use shape prop, and it should happen before the sharding prop here, because otherwise the view op propagation would fail, and I think we shouldn't worry about recovering the shape in sharding prop, because by the time we enter sharding prop, we already finish sharding prop for a graph with meta tensor